mesonbuild/dependencies/dev.py

```diff
@@ -546,6 +549,15 @@ def __init__(self, environment: 'Environment', kwargs: JNISystemDependencyKW):
         self.java_home = environment.properties[self.for_machine].get_java_home()
         if not self.java_home:
             self.java_home = pathlib.Path(shutil.which(self.javac.exelist[0])).resolve().parents[1]
+            if m.is_darwin():
+                problem_java_prefix = pathlib.Path('/System/Library/Frameworks/JavaVM.framework/Versions')
+                if problem_java_prefix in self.java_home.parents:
+                    res = subprocess.run(['/usr/libexec/java_home', '--failfast', '--arch', m_cpu_family],
```

Comment:
@eli-schwartz any thoughts on what if anything should be logged if this command fails?

Reply:
"Your binary is not an allowed client of /usr/libexec/java_home", and leave it to the user to guess whether the binary is javac or Meson.

We want macOS users to feel at home, after all. Plus it's good to follow Apple's UX guidelines.